### PR TITLE
Improve docs in taproot module

### PR DESCRIPTION
**Pull request by tcharding**
I should have done this PR a month ago, my bad. This one is kind of important IMO because we are going to have so many people looking at this part of the code soon as we release.

As has been done in other places in the codebase; improve the docs in the `taproot` module by doing:

- Use full sentences (capital letters + full stops)
- Use back ticks and links for types where appropriate
- Fix grammar
- Fix stale docs
- Use third person for describing functions
- Use 100 character line width
- Use markdown sections (`# Examples`, `# Returns`) where appropriate
- Separate brief heading from extended description when appropriate
- Use `///` for all functions/types (both private and public)

I also did:

- Build the docs and check all the links
- Read all the built docs, check for sanity and pretty-ness

Its all in one patch, I couldn't really tease it apart. I can try a bit harder if it proves too annoying to review.

### CODE REVIEW

**Comment by tcharding** [comment]
CC @sanket1729 since this is taproot stuff.

File: src/util/taproot.rs
Comment (sanket1729) [inline src/util/taproot.rs:714]: This is meant to `Self::serialize().len()` instead of `Self::len()`

**Review by sanket1729 — COMMENTED** [review]
Thank you so much for doing this.
Left one comment that needs to be changed. I need to set my editor to make use of 100 characters whenever available while commenting. I promise I will be more conscious while writing docs for my subsequent PRs.

File: src/util/taproot.rs
Comment (apoelstra) [inline src/util/taproot.rs:692]: In f488093a04a02573507fcb8bdbd35b4bed7ba7b2:

Type `TaprootErrro`

File: src/util/taproot.rs
Comment (apoelstra) [inline src/util/taproot.rs:745]: In f488093a04a02573507fcb8bdbd35b4bed7ba7b2:

This comma should probably be a full stop.

**Comment by tcharding** [comment]
> I promise I will be more conscious while writing docs for my subsequent PRs.

I'm a firm believer that folk should work on what stimulates them, if writing clean docs is not your thing then just stick to doing you thing well :) I do these docs fixes at the end of the days or when I'm brain fried, no stress.

File: src/util/taproot.rs
Comment (tcharding) [reply src/util/taproot.rs:714]: Of course, thanks.

File: src/util/taproot.rs
Comment (tcharding) [reply src/util/taproot.rs]: Done, thanks.

File: src/util/taproot.rs
Comment (tcharding) [reply src/util/taproot.rs]: Nice, thanks.

**Comment by tcharding** [comment]
Force push includes:
- the minor fixes suggested in review as part of patch 1 
- the removal of bip341 docs as a separate patch
- I've rudely kept in my 'improved' version of the docs on `with_huffman_tree` even though @dr-orlovsky  already fixed the stale types thing (there has to be some perks to being the monkey that does all these docs fixes :)

**Comment by tcharding** [comment]
This PR is a candidate for either the 'RC-fix' label or the 0.28 milestone, leaving for another maintainer to add so as not to be too pushy with my own changes.

**Review by sanket1729 — APPROVED** [review]
ACK c25eddd18706bd4fcdad205db804b2af32d815df

**Comment by sanket1729** [comment]
If we get two reviews in time, sure. I don't mind merging any changes to documentation before release.

File: src/util/taproot.rs
Comment (dr-orlovsky) [inline src/util/taproot.rs:214]: nit:
```suggestion
    /// the `merkle_root` if there is no script path.
```

File: src/util/taproot.rs
Comment (dr-orlovsky) [inline src/util/taproot.rs:134]: nit, but not sure:
```suggestion
/// Maximum depth of a taproot tree script spending path.
```

**Review by dr-orlovsky — APPROVED** [review]
ACK c25eddd18706bd4fcdad205db804b2af32d815df

Will wait for @apoelstra review before merge, since he had done previous review.

All my suggestions are nits, which may be addressed in the follow up PR. Not sure about the articles, but I assume that all function arg references should be prefixed with `the`.

File: src/util/taproot.rs
Comment (apoelstra) [reply src/util/taproot.rs:475]: I think the existing grammar is correct.

**Review by apoelstra — APPROVED** [review]
ACK c25eddd18706bd4fcdad205db804b2af32d815df

**Comment by tcharding** [comment]
Thanks for the review @dr-orlovsky, I'll see to your nits.

File: src/util/taproot.rs
Comment (tcharding) [reply src/util/taproot.rs:523]: Used with some minor tweaks.

File: src/util/taproot.rs
Comment (tcharding) [reply src/util/taproot.rs:434]: I'll try to explain why the original is, IMO, correct. 'depth' is not a noun in this sentence, conceptually its a location in the tree, in other words 'depth' could be replaced by 'the layer of the tree represented by usize' - hence we use 'depth' without using 'the'. Does that make sense?

File: src/util/taproot.rs
Comment (tcharding) [reply src/util/taproot.rs:134]: I believe 'key spend' and 'script spend' are the terms being used across the eccosystem, I could be wrong.

**Comment by dr-orlovsky** [comment]
@apoelstra @tcharding sorry for the mess with the articles in my suggestions, will not do that in a (the?) future. As a non-native  speaker the whole concept of articles for me (with native Slavic language background) is ungraspable. I thought I did understood the logic of using them, but in fact not :)

**Comment by apoelstra** [comment]
No worries @dr-orlovsky. I am a native English speaker and I do not know any real rules for how articles are used. (If it helps I have the same difficulty in French, e.g. I was surprised that [this book title](https://nitter.it/pic/enc/bWVkaWEvRlBHZDZUYWFBQVV2S2hsLnBuZz9uYW1lPW9yaWc=) uses both `les` and `des` as articles in places where the English title has no articles at all.)

**Pull request merged**
